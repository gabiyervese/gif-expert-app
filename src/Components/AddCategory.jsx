import { useState } from "react";

export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState("");

  const onInputChange = (event) => {
    //En el evento on change para los inputs con los que el user interactua siempre evaluare el "set" con el event.target.value, que es lo que el usuario introduce, de esta forma al crear la funcion por fuera, abajo en el onchange puedo colocar solo el nombre de la funcion sin necesidad de hacer toda la operacion abajo.
    setInputValue(event.target.value);
  };

  const onSubmit = (event) => {
    //con el prevent evito que cada vez que enviemos el form se refresque la pagina entera lo cual hace por defecto.
    event.preventDefault();
    const newInputValue = inputValue.trim();
    //Con este if condiciomanos a que no se permita elviar el valor del input si su longitud es <= a un caracter.
    if (newInputValue.length <= 1) return;
    //Ejecuto la funcion con el newInputValue
    onNewCategory(newInputValue);
    //Con esto al enviar el form el input vuelve a su estado vacio.
    setInputValue(" ");
  };

  return (
    <form onSubmit={onSubmit}>
      <input
        type="text"
        placeholder="Search gifs"
        // No debo olvidar colocar el valor del input!!!
        value={inputValue}
        onChange={onInputChange}
      />
    </form>
  );
};
