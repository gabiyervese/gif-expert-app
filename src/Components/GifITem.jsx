import propTypes from "prop-types";

export const GifItem = ({ title, url }) => {
  return (
    <>
      <h4>{title}</h4>
      <img src={url} alt={title} />
    </>
  );
};

GifItem.propTypes = {
  title: propTypes.string.isRequired,
  url: propTypes.string.isRequired,
};
