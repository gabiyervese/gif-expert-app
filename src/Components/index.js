//Este es mi archivo de barril

export * from "./AddCategory";
export * from "./GifGrid";
export * from "./GifITem";
