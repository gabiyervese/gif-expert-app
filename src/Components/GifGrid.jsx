import { GifItem } from "./GifITem";
import { useFetchGifs } from "../Hooks/useFetchGifs";
import propTypes from "prop-types";

export const GifGrid = ({ category }) => {
  //Aqui implementamos el custom hook para limpiar el componente de tanto codigo.
  const { images, isLoading } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>
      {/* Si isLoading es true coloca el elemento */}
      {isLoading && <h2>Cargando...</h2>}
      <ul className="card-grid">
        {images.map(({ url, title, id }) => {
          return (
            <li key={id} className="card">
              <GifItem url={url} title={title} />;
              {/* Si quisiera esparcir todas las props en el componente puedo usar {...objeto} de esta forma es muy util cual quiero enviar tooodas las props y son muchas */}
            </li>
          );
        })}
      </ul>
    </>
  );
};

GifGrid.propTypes = {
  category: propTypes.string.isRequired,
};
