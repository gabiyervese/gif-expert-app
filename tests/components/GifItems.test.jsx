import { render, screen } from "@testing-library/react";
import { GifItem } from "../../src/Components/GifITem";

describe("Pruebas en <GifItem/>", () => {
  const title = "one piece";
  const url = "http://onepiece.com/monkeyDluffy";

  //Test #1
  test("Debe hacer match con el snapshot ", () => {
    const { container } = render(<GifItem title={title} url={url} />);
    expect(container).toMatchSnapshot();
  });

  //Test #2
  test("Debe mostrar la imagen con el URL y el ALT indicado ", () => {
    render(<GifItem title={title} url={url} />);
    // screen.debug();

    // expect(screen.getByRole("img").src).toBe(url);
    // expect(screen.getByRole("img").alt).toBe(title);

    //Esta es una forma mas limpia de hacer lo de arriba:
    const { src, alt } = screen.getByRole("img");
    expect(src).toBe(url);
    expect(alt).toBe(title);
  });

  //Test #3
  test("Debe mostrar el title como texto en el componente", () => {
    render(<GifItem title={title} url={url} />);
    //Aqui buscamos que el title efectivamente exista.
    expect(screen.getByText(title)).toBeTruthy();
  });
});
