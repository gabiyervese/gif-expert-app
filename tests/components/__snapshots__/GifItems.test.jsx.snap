// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifItem/> Debe hacer match con el snapshot  1`] = `
<div>
  <h4>
    one piece
  </h4>
  <img
    alt="one piece"
    src="http://onepiece.com/monkeyDluffy"
  />
</div>
`;
